### PR TITLE
forward encoding for single field struct

### PR DESCRIPTION
**Pull request by thiolliere**
Fix https://github.com/paritytech/parity-codec/issues/43

### CODE REVIEW

File: derive/src/encode.rs [hunk 1]
```diff
@@ -79,9 +119,51 @@ fn encode_fields<F>(
...
+		return quote! {
+			fn using_encoded<R, F: FnOnce(&[u8]) -> R>(&self, #closure: F) -> R {
+				#optimisation
+			}
+		};
+	}
+
+	let self_ = quote!(self);
+	let ref dest = quote!(dest);
+	let encoding = match *data {
```
Comment (bkchr) [inline derive/src/encode.rs]: Maybe we could move the code below into an own function as well.

File: derive/src/encode.rs [hunk 2]
```diff
@@ -79,9 +119,51 @@ fn encode_fields<F>(
...
+pub fn quote(data: &Data, type_name: &Ident) -> TokenStream {
 	let call_site = Span::call_site();
-	match *data {
+
+	// optimisation for single field struct
+	let ref closure = quote!(f);
+	let optimisation = match *data {
+		Data::Struct(ref data) => {
+			match data.fields {
+				Fields::Named(ref fields) => if fields.named.len() == 1 {
```
[the remark this answers is not in the document]
Comment (thiolliere) [reply derive/src/encode.rs]: this is so cool!

**Comment by thiolliere** [comment]
seems good to me now, I cleaned quote_spanned in older code as well.

File: derive/src/encode.rs [hunk 3]
```diff
@@ -177,8 +249,22 @@ pub fn quote(data: &Data, type_name: &Ident, self_: &TokenStream, dest: &TokenSt
...
+	quote! {
+		fn encode_to<EncOut: _parity_codec::Output>(&#self_, #dest: &mut EncOut) {
+			#encoding
+		}
 	}
 }
+
+pub fn quote(data: &Data, type_name: &Ident) -> TokenStream {
+	if let Some(implementation) = try_impl_encode_single_field_optimisation(data) {
+		return implementation;
```
Comment (bkchr) [inline derive/src/encode.rs]:
```suggestion
		implementation
```

File: derive/src/encode.rs [hunk 4]
```diff
@@ -177,8 +249,22 @@ pub fn quote(data: &Data, type_name: &Ident, self_: &TokenStream, dest: &TokenSt
...
+	}
```
Comment (bkchr) [inline derive/src/encode.rs]:
```suggestion
	} else {
```

File: derive/src/encode.rs [hunk 5]
```diff
@@ -177,8 +249,22 @@ pub fn quote(data: &Data, type_name: &Ident, self_: &TokenStream, dest: &TokenSt
...
+	impl_encode(data, type_name)
```
Comment (bkchr) [inline derive/src/encode.rs]:
```suggestion
		impl_encode(data, type_name)
```

File: derive/src/encode.rs [hunk 6]
```diff
@@ -177,8 +249,22 @@ pub fn quote(data: &Data, type_name: &Ident, self_: &TokenStream, dest: &TokenSt
...
+}
```
Comment (bkchr) [inline derive/src/encode.rs]:
```suggestion
	}
}
```

**Pull request merged**